model.py

```diff
@@ -1,103 +1,1222 @@
+        The parameter objects that correspond to this node. Single Param is
+        also accepted.
+      parent; Model, optional:
+        The parent of this node.
+      children; list(Model), or Model:
+        The children of this node. Single Model is also accepted
+    '''
+
+    # Set the parameter names for this layer
+    node_par_names = ('q', 'dphi', 'rwd')
+
+    def ln_prior(self, verbose=False):
+        '''Before we calculate the ln_prior of myself or my children, I check
+        that my parameters are valid. I check that dphi is not too large for
+        the current value of q.
+
+        If other constraints on the core parameters become necessary, they
+        should go here. If these tests fail, -np.inf is immediately returned.
+        '''
+        lnp = 0.0
+
+        # Check that dphi is within limits
+        tol = 1e-6
+
+        dphi = getattr(self, 'dphi').currVal
+        q = getattr(self, 'q').currVal
+
+        try:
+            # Get the value of dphi that we WOULD have at an
+            # inclination of 90 degrees
+            maxphi = roche.findphi(q, 90.0)
+
+            # If dphi is out of range, return negative inf.
+            if dphi > (maxphi - tol):
+                if verbose:
+                    msg = "{} has a dphi out of tolerance!\nq: {:.3f}"
+                    msg += "\ndphi: {:.3f}, max: {:.3f} - {:.3g}"
+                    msg += "\nReturning inf.\n\n"
+
+                    msg.format(self.name, q, dphi, maxphi, tol)
+
+                    print(msg)
+                return -np.inf
+
+        except:
+            # If we get here, then roche couldn't find a dphi for this q.
+            # That's bad!
+            if verbose:
+                msg = "Failed to calculate a value of dphi at node {}"
+                print(msg.format(self.name))
+            return -np.inf
+
+        # Then, if we pass this, move on to the 'normal' ln_prior calculation.
+        lnp += super().ln_prior(verbose=verbose)
+        return lnp
+
+
+class GPEclipse(Eclipse):
+    '''This is a subclass of the Eclipse class. It uses the Gaussian Process.
+
+    This version will, rather than evaluating via chisq, evaluate the
+    likelihood of the model by calculating the residuals between the model
+    and the data and computing the likelihood of those data given certain
+    Gaussian Process hyper-parameters. '''
+
+    # Set the initial values of q, rwd, and dphi. These will be used to
+    # caclulate the location of the GP changepoints. Setting to initially
+    # unrealistically high values will ensure that the first time
+    # calcChangepoints is called, the changepoints are calculated.
+    _olddphi = 9e99
+    _oldq = 9e99
+    _oldrwd = 9e99
+
+    # _dist_cp is initially set to whatever, it will be overwritten anyway.
+    _dist_cp = 9e99
+
+    def __init__(self, ampin_GP, ampout_GP, tau_GP, *args, **kwargs):
+
+        # GP hyperparameters
+        self.ampin = ampin_GP
```

Comment:
Shouldn't these be model parameters in the same way as all others? E.g.

```python
class GPEclipse(Eclipse):
    node_par_names = (..., 'ampin', 'ampout', 'tau')
    ...
```

Reply:
You're right -- dumb oversight!

Changed this, but GP are still largely untested until I get the rest of it streamlined.